src/object/cli_obj.c

```diff
@@ -5277,36 +5308,69 @@ obj_update_shards_get(struct dc_object *obj, daos_obj_fetch_t *args, unsigned in
 {
 	uint8_t		*tgt_bitmap;
 	uint32_t	failure_cnt = 0;
+	int		grp_idx;
+	uint32_t	grp_start;
+	uint32_t	shard_nr = 0;
 	int		i;
-	int		rc;
+	int		rc = 0;
 
-	rc = obj_dkey2grpmemb(obj, obj_auxi->dkey_hash, map_ver, shard, shard_cnt);
-	if (rc || !obj_is_ec(obj))
-		return rc;
+	if (!obj_is_ec(obj))
+		return obj_dkey2grpmemb(obj, obj_auxi->dkey_hash, map_ver, shard, shard_cnt);
 
-	tgt_bitmap = obj_auxi->reasb_req.tgt_bitmap;
-	for (i = 0; i < *shard_cnt; i++) {
-		int shard_idx;
+	grp_idx = obj_dkey2grpidx(obj, obj_auxi->dkey_hash, map_ver);
+	if (grp_idx < 0)
+		return grp_idx;
 
-		if (isclr(tgt_bitmap, i))
-			continue;
+	grp_start = grp_idx * obj_get_grp_size(obj);
+	tgt_bitmap = obj_auxi->reasb_req.tgt_bitmap;
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+	for (i = 0; i < obj_get_grp_size(obj); i++) {
+		struct dc_obj_shard	*obj_shard;
+		unsigned int		shard_id;
+		int			shard_idx;
 
-		shard_idx = *shard + i;
+		shard_idx = grp_start + i;
 		D_ASSERTF(shard_idx < obj->cob_shards_nr, "%d >= %u\n",
 			  shard_idx, obj->cob_shards_nr);
 
-		if (obj_shard_is_invalid(obj, shard_idx, DAOS_OBJ_RPC_UPDATE) ||
+		obj_shard = &obj->cob_shards->do_shards[shard_idx];
+		if (obj_shard->do_target_id == -1 || obj_shard->do_shard == -1 ||
 		    unlikely(DAOS_FAIL_CHECK(DAOS_FAIL_SHARD_NONEXIST))) {
 			if (++failure_cnt > obj_ec_parity_tgt_nr(obj_get_oca(obj))) {
 				D_ERROR(DF_OID" failures %u is more than parity cnt.\n",
 					DP_OID(obj->cob_md.omd_id), failure_cnt);
+				D_RWLOCK_UNLOCK(&obj->cob_lock);
 				D_GOTO(out, rc = -DER_IO);
 			}
+
 			D_DEBUG(DB_IO, DF_OID" skip shard %d\n", DP_OID(obj->cob_md.omd_id),
 				shard_idx);
-			clrbit(tgt_bitmap, i);
+			if (obj_shard->do_shard != -1)
```

Comment:
just confirm why need check "if (obj_shard->do_shard != -1)" here? is it possible "obj_shard->do_target_id == -1 && obj_shard->do_shard != -1"? (which case?)

Reply:
oh, since the following line only use do_shard, so I only check do_shard here.  

Hmm, I do not think there are cases that only one is -1, actually I though do_shard should never be -1, only do_target_id can be -1 if no target available, though not fixing it in this patch.

Reply:
ok, then need not check "if (obj_shard->do_shard != -1)" here to avoid confuse, thx